### PR TITLE
Share `:build-cache-packaging`

### PR DESCRIPTION
**Pull request by lptr**
Fixes https://github.com/gradle/gradle-private/issues/2417

## TODO

- [ ] ~~Remove dependency gradle-native by replacing FileSystem~~ -- see https://github.com/gradle/gradle/pull/11821
- [x] Create build-cache-base so packaging can depend on it
	- OriginMetadata + Reader/Writer
- [x] Make `BuildCacheCommandFactory` into an interface and move it to `:build-cache`
- [x] Move `DefaultBuildCacheCommandFactory` to :core
- [x] ~~Move `BuildCacheKey` to `:build-cache-base`~~ not necessary after moving the `BuildCacheCommandFactory` out of `:build-cache-packaging`
- [x] Move `@NonNullApi` to `:pineapple` and replace all usage of `@NonNullInternalApi` with it – see https://github.com/gradle/gradle/pull/11820
- [x] Remove dependency on `FilePathUtil`

### CODE REVIEW

**Comment by lptr** [comment]
This PR now removes all remaining dependencies of `:build-cache-packaging` except for the ones removed by #11821. Let's do a review and merge it asap.

File: subprojects/build-cache-base/build-cache-base.gradle.kts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 the original author or authors.
+ * Copyright 2017 the original author or authors.
```
Comment (britter) [inline subprojects/build-cache-base/build-cache-base.gradle.kts]: Why this change? Shouldn't that be 2020?

File: subprojects/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java [hunk 2]
```diff
@@ -16,15 +16,13 @@
...
-import com.google.common.base.Preconditions;
-
 public class OriginMetadata {
 
     private final String buildInvocationId;
     private final long executionTime;
 
     public OriginMetadata(String buildInvocationId, long executionTime) {
-        this.buildInvocationId = Preconditions.checkNotNull(buildInvocationId, "buildInvocationId");
+        this.buildInvocationId = buildInvocationId;
```
Comment (britter) [inline subprojects/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java]: ⭕️ You can use `Objects.requireNonNull()` if you want to keep this assertion (which I think is a good thing for data objects)

File: subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/BuildCacheCommandFactory.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import org.gradle.caching.BuildCacheKey;
+import org.gradle.caching.internal.CacheableEntity;
+import org.gradle.caching.internal.origin.OriginMetadata;
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
+import org.gradle.internal.snapshot.FileSystemSnapshot;
+
+import java.util.Map;
+
+public interface BuildCacheCommandFactory {
+    BuildCacheLoadCommand<LoadMetadata> createLoad(BuildCacheKey cacheKey, CacheableEntity entity);
```
Comment (britter) [inline subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/BuildCacheCommandFactory.java]: ⭕️ This looks a little bit different on ower site, I need to figure out why.

File: subprojects/files/src/main/java/org/gradle/internal/file/FilePathUtil.java [hunk 4]
```diff
@@ -14,21 +14,28 @@
...
+    private static final String[] EMPTY_STRING_ARRAY = {};
     // On Windows, / and \ are separators, on Unix only / is a separator.
     private static final String FILE_PATH_SEPARATORS = File.separatorChar != '/' ? ("/" + File.separator) : File.separator;
 
     private FilePathUtil() {
     }
 
     public static String[] getPathSegments(String path) {
-        return StringUtils.split(path, FILE_PATH_SEPARATORS);
+        StringTokenizer tokenizer = new StringTokenizer(path, FILE_PATH_SEPARATORS);
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/files/src/main/java/org/gradle/internal/file/FilePathUtil.java]: Why is it supposedly slower? Before making the change I checked the code in `StringUtils` and it doesn't seem to be doing anything special to make it fast. It adds stuff to a `List`, same as what we do now with `StringTokenizer`.

[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/files/src/main/java/org/gradle/internal/file/FilePathUtil.java]: The comment in `StringUtils` claims it:

```
        // Performance tuned for 2.0 (JDK1.4)
        // Direct code is quicker than StringTokenizer.
```

It doesn't use `StringTokenizer`.

Still, the comment doesn't need to be correct anymore. I wanted to see performance test results for the change.

File: subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/BuildCacheCommandFactory.java (hunk 3 above)
Comment (lptr) [reply subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/BuildCacheCommandFactory.java]: This is fixed on your side now, right?

File: subprojects/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java (hunk 2 above)
Comment (lptr) [reply subprojects/build-cache-base/src/main/java/org/gradle/caching/internal/origin/OriginMetadata.java]: I know, but we don't actually check anywhere on internal APIs for `null` parameters. I'm also not a big fan of defensive code like this, especially since we have nullability checks now in IntelliJ.

File: subprojects/build-cache-base/build-cache-base.gradle.kts (hunk 1 above)
Comment (lptr) [reply subprojects/build-cache-base/build-cache-base.gradle.kts]: I have no idea how we ended up with this number. Updated to `2020`.

File: subprojects/files/src/main/java/org/gradle/internal/file/FilePathUtil.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/files/src/main/java/org/gradle/internal/file/FilePathUtil.java]: I'll run some performance tests.

File: subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/BuildCacheCommandFactory.java (hunk 3 above)
Comment (britter) [reply subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/BuildCacheCommandFactory.java]: Yes, I'll merge the PR this morning.

**Comment by lptr** [comment]
Damn, looks like we do more GC with the changes, I'll take a look at that splitter...

**Comment by lptr** [comment]
> Damn, looks like we do more GC with the changes, I'll take a look at that splitter... https://builds.gradle.org/repository/download/Gradle_Check_PerformanceTestCoordinator/30146906:id/report-performance-performance-tests.zip%21/report/tests/executing-tasks-with-classloader-isolation-work%3D10---workers%3D1.html?redirectSupported=false

As discussed with @wolfs that performance test is actually flaky, so we'll just go ahead with merging this as soon as `master` is green again.
